dali/kernels/shape.h

```diff
@@ -0,0 +1,419 @@
+  using Base = TensorShapeBase<std::vector<int64_t>>;
+  TensorShape(const std::vector<int64_t> &s) : Base(s) {}
+
+  template <size_t N>
+  TensorShape(std::array<int64_t, N> s) : Base(typename Base::container_type(s.begin(), s.end())) {}
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : Base(typename Base::container_type{i0, int64_t{s}...}) {}
+
+  TensorShape() = default;
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <int other_ndim>
+  TensorShape(const TensorShape<other_ndim> &other)
+      : Base(typename Base::container_type(other.shape.begin(), other.shape.end())) {}
+
+  template <int other_ndim>
+  TensorShape &operator=(const TensorShape<other_ndim> &other) {
+    shape = Base::container_type(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  // We allow only explicit conversion from dynamic to static dim
+  template <int other_ndim>
+  TensorShape<other_ndim> to_static() const {
+    static_assert(other_ndim != DynamicDimensions,
+                  "Conversion to static only allowed for static shape");
+    assert(size() == other_ndim);
+    TensorShape<other_ndim> shape;
+    for (int i = 0; i < other_ndim; i++) {
+      shape[i] = (*this)[i];
+    }
+    return shape;
+  }
+
+  TensorShape<DynamicDimensions> first(int count) {
+    assert(0 <= count && count <= size() &&
+           "Number of elements in subshape must be between 0 and size()");
+    TensorShape<DynamicDimensions> result;
+    result.shape.resize(count);
+    for (int i = 0; i < count; i++) {
+      result[i] = (*this)[i];
+    }
+    return result;
+  }
+
+  TensorShape<DynamicDimensions> last(int count) {
+    assert(0 <= count && count <= size() &&
+           "Number of elements in subshape must be between 0 and size()");
+    TensorShape<DynamicDimensions> result;
+    result.shape.resize(count);
+    int start_offset = size() - count;
+    for (int i = 0; i < count; i++) {
+      result[i] = (*this)[start_offset + i];
+    }
+    return result;
+  }
+};
+
+template <int ndim>
+struct TensorShape : public TensorShapeBase<std::array<int64_t, ndim>> {
+  using Base = TensorShapeBase<std::array<int64_t, ndim>>;
+  TensorShape(const std::array<int64_t, ndim> &s) : Base(s) {}
+  // We just want zero-initialization
+  TensorShape() = default;
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : Base(typename Base::container_type{i0, int64_t{s}...}) {
+    static_assert(sizeof...(Ts) == ndim - 1, "Number of shapes passed must match ndim");
+  }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> first() {
```

Comment:
Dynamic first and last would also be nice to have.

Reply:
Done